lib/streamlit/runtime.py

```diff
@@ -0,0 +1,591 @@
+
+            # Store the eventloop we're running on so that we can schedule
+            # callbacks on it when necessary. (We can't just call
+            # `asyncio.get_running_loop()` whenever we like, because we have
+            # some functions, e.g. `stop`, that can be called from other
+            # threads, and `asyncio.get_running_loop()` is thread-specific.)
+            self._eventloop = asyncio.get_running_loop()
+
+            if on_started is not None:
+                on_started()
+
+            while not self._must_stop.is_set():
+                if self._state == RuntimeState.WAITING_FOR_FIRST_SESSION:
+                    await asyncio.wait(
+                        [self._must_stop.wait(), self._has_connection.wait()],
+                        return_when=asyncio.FIRST_COMPLETED,
+                    )
+
+                elif self._state == RuntimeState.ONE_OR_MORE_SESSIONS_CONNECTED:
+                    self._need_send_data.clear()
+
+                    # Shallow-clone our sessions into a list, so we can iterate
+                    # over it and not worry about whether it's being changed
+                    # outside this coroutine.
+                    session_infos = list(self._session_info_by_id.values())
+
+                    for session_info in session_infos:
+                        msg_list = session_info.session.flush_browser_queue()
+                        for msg in msg_list:
+                            try:
+                                self._send_message(session_info, msg)
+                            except SessionClientDisconnectedError:
+                                self.close_session(session_info.session.id)
+
+                            # Yield for a tick after sending a message.
+                            await asyncio.sleep(0)
+
+                    # Yield for a few milliseconds between session message
+                    # flushing.
+                    await asyncio.sleep(0.01)
+
+                elif self._state == RuntimeState.NO_SESSIONS_CONNECTED:
+                    await asyncio.wait(
+                        [self._must_stop.wait(), self._has_connection.wait()],
+                        return_when=asyncio.FIRST_COMPLETED,
+                    )
+
+                else:
+                    # Break out of the thread loop if we encounter any other state.
+                    break
+
+                await asyncio.wait(
+                    [self._must_stop.wait(), self._need_send_data.wait()],
+                    return_when=asyncio.FIRST_COMPLETED,
+                )
+
+            # Shut down all AppSessions
+            for session_info in list(self._session_info_by_id.values()):
+                session_info.session.shutdown()
+
+            self._set_state(RuntimeState.STOPPED)
+
+        except Exception:
+            traceback.print_exc()
+            LOGGER.info(
+                """
+Please report this bug at https://github.com/streamlit/streamlit/issues.
+"""
```

Comment:
nit: not particularly important, but I think this message is short enough that it isn't worth writing in a docstring (I'm guessing at one point it was longer, so the docstring was necessary)

Reply:
I think what this is actually doing is adding an additional newline, so that the request stands out below the stack trace that will be printed right above it.